File: 大话数据结构原书C版以及Java版翻译/图/Kruskal.java

import java.util.Arrays;
import java.util.Scanner;

/**Kruskal算法 内部随便用的冒泡排序 如果以后有机会再改进 -最小生成树
 * Created by han on 2016/12/24.
 */
public class Kruskal {


    public    final  int MAXVEX=100;
    public    final  int MAXSIZE=5;
    public    final  int MAXEDGE=20;
    public    final  int MAX=100;
    public    final  int INFINITY=65535;

    //  邻接矩阵

    class MGraph{
        String[] vexs=new String[MAXVEX];
        int[][] arc = new int[MAXVEX][MAXVEX];
        int numVertexes,numEdges;

        @Override
        public String toString() {
            return "MGraph{" +
                    "vexs=" + Arrays.toString(vexs) +
                    ", arc=" + Arrays.toString(arc) +
                    ", numVertexes=" + numVertexes +
                    ", numEdges=" + numEdges +
                    '}';
        }
    }

    class Edge{
        int begin;
        int end;
        int weigth;

        @Override
        public String toString() {
            return "Edge{" +
                    "begin=" + begin +
                    ", end=" + end +
                    ", weigth=" + weigth +
                    '}';
        }
    }

    void CreateMGrapht(MGraph G){

        int i,j,k,w;

        System.out.println("输入顶点数和边数");
        Scanner sc = new Scanner(System.in);
        G.numVertexes = sc.nextInt();
        G.numEdges = sc.nextInt();
        System.out.println("输入顶点信息");
        for (i=0;i<G.numVertexes;i++)
            G.vexs[i] = sc.next();
        for (i=0;i<G.numVertexes;i++)
            for (j=0;j<G.numVertexes;j++)
                G.arc[i][j]=INFINITY;

        for (k=0;k<G.numEdges;k++){
            System.out.println("输入边(vi,vj)上的下标i,上标j和权w");
            i = sc.nextInt();
            j = sc.nextInt();
            w=sc.nextInt();
            G.arc[i][j]=w;
//            G.arc[j][i]=G.arc[i][j];
        }
    }


    void MiniSpanTree_Kruskal(MGraph G){
        int i,n,m;
        Edge[] edges = new Edge[MAXEDGE];
        int[] parent = new int[MAXEDGE];
        int min=INFINITY;
        int k=0,j;
        Edge middleEdge;

        for (j=0;j<G.numVertexes;j++){
            for (i=0;i<G.numVertexes;i++){
                if(G.arc[i][j] <min) {
                    edges[k] = new Edge();
                    edges[k].begin = i;
                    edges[k].end = j;
                    edges[k].weigth = G.arc[i][j];
                    k++;
                }
            }
        }

        for (i=0;i<k;i++){
            for (j=i+1;j<k;j++){
                if(edges[i].weigth>edges[j].weigth){
                    middleEdge=edges[i];
                    edges[i]=edges[j];
                    edges[j]=middleEdge;
                }
            }
        }

       for(i=0;i<G.numVertexes;i++)
           parent[i]=0;
        for (i=0;i<G.numVertexes;i++){
            n=Find(parent,edges[i].begin);
            m=Find(parent,edges[i].end);
            if(n!=m){
                parent[n]=m;
                System.out.println("("+edges[i].begin+","+edges[i].end+")"+edges[i].weigth);
            }
        }

        }


    int Find(int[] parent,int f){
        while (parent[f]>0)
            f=parent[f];
        return f;
    }

    public static void main(String[] args) {
        MGraph G = new Kruskal().new MGraph();
        Kruskal kruskal = new Kruskal();
        kruskal.CreateMGrapht(G);
        kruskal.MiniSpanTree_Kruskal(G);

    }

}

File: 大话数据结构原书C版以及Java版翻译/图/Kruskal.c
#include <stdio.h> //Kruskal算法 -最小生成树
#define OK 1
#define ERROR 0
#define TRUE 1
#define FALSE 0
#define MAXSIZE 100
#define MAXEDGE 20
#define MAXVEX 100  /*最大顶点数 应由用户定义*/
#define INFINITY 65535 /*用65535表示OO*/
typedef char VertexType; /*顶点类型应由用户定义*/
typedef int  EdgeType; /*边上的权值类型应由用户定义*/
typedef int QElemType;
#define    MAX 100
typedef  int  Boolean;/*Boolean是布尔类型 其值是True或False*/
typedef int Status;
typedef struct {
	VertexType vexs[MAXVEX];/*顶点表*/
	EdgeType arc[MAXVEX][MAXVEX];/*临接矩阵 可看成表*/
	int numVertexes,numEdges;/*图中当前的顶点数和边数*/
} MGraph;


 /*对边集数组*/
typedef struct
{
	int begin;
	int end;
	int weight;
}Edge;

/*Kruskal算法生成最小生成树*/
void MiniSpanTree_Kruskal(MGraph G)//生成最小生成树
{
	int i,n,m;
	Edge edges[MAXEDGE];//定义边集数组
	int parent[MAXEDGE];//定义一数组用来判断边与边是否形成环路
	//此处省略将邻接矩阵G转换为边集数组edges并按权从小到大排序的代码
	int min=INFINITY;
	int k,j;
	Edges middleEdge;

	for(j=0;j<G.numVertexes;j++){
    	for(i=0;i<G.numVertexes;i++){
	        if(G.arc[i][j] <min){
	            edges[k].begin=i;
	            edges[k].end=j;
	            edges[k].weight=G.arc[i][j];
				k++;
			}
	   }
}
//冒泡排序
  for(i=0;i<k;i++){
    	for(j=i+1;j<k;j++){
	       if(edges[i].weight>edges[j].weight){
	       	  middleEdge=edges[i];
	       	  edges[i]=edges[j];
	       	  edges[j]=middleEdge;
		   }
		}

  }

	for(i=0;i<G.numEdges;i++)
	  parent[i]=0;//初始化数组值为0
	 for(i=0;i<G.numEdges;i++)//循环每一条边
	 {
	 	n=Find(parent,edges[i].begin);
	 	m=Find(parent,edges[i].end);
	 	if(n!=m)//假如n与m不等 说明此边没有与现有生成树形成环路
		 {
		 	parent[n]=m;//将此边的结尾顶点放入下标为顶点的parent中
			 //表示此顶点已经在生成树集合中
		 	printf("(%d,%d) %d",edges[i].begin
			 ,edges[i].end,edges[i].weight);
		  }
	  }
 }

 void Fine(int *parent,int f)//查找连线结点的尾部下标
 {
 	while(parent[f] >0)
 	f=parent[f];
 	return f;
 }
